feat: StreamCall signature, video client creation

szuperaz opened the pull request:
## React SDK
- `StreamCall` now only accepts a `Call` instance, integrators have to take care of join/getOrCreate
- `useStreamVideoClient` has been removed
- Docs will be updated in separate PR

## RN SDK
- `useStreamVideoClient` moved from `react-bindings` to RN SDK

oliverlaz commented on packages/client/src/StreamVideoClient.ts:
```diff
@@ -69,11 +72,19 @@ export class StreamVideoClient {
-      user,
-      tokenOrProvider,
-    );
+    const connectUser = () => {
+      return this.streamClient.connectUser(
+        // @ts-expect-error
+        user,
+        tokenOrProvider,
+      );
+    };
+    this.connectionPromise = this.disconnectionPromise
+      ? this.disconnectionPromise.then(() => connectUser())
+      : connectUser();
+
+    this.connectionPromise?.finally(() => (this.connectionPromise = undefined));
+    const connectUserResponse = await this.connectionPromise;
```
Do we need to handle this here? I'd probably leave this up to the integrators... In our hook, we had to deal with this because of React's StrictMode where every effect is executed twice in dev mode:
- https://react.dev/reference/react/StrictMode

oliverlaz commented on sample-apps/react/egress-composite/src/CompositeApp.tsx:
```diff
@@ -19,14 +19,22 @@ export const CompositeApp = () => {
+    return () => {
+      client.disconnectUser();
+    };
+  }, [client, config.token]);
```
```suggestion
  }, [client, config.userId, config.token]);
```

I think `userId` should be provided as a dependency here.

oliverlaz commented on sample-apps/react/livestream-app/src/hosts/Hosts.tsx:
```diff
@@ -24,28 +21,37 @@ export const characters = [
-  });
+    };
+
+    client.connectUser(user, tokenProvider);
```
```suggestion
    client.connectUser(user, tokenProvider).catch(console.error);
```

oliverlaz commented on sample-apps/react/livestream-app/src/hosts/Hosts.tsx:
```diff
@@ -24,28 +21,37 @@ export const characters = [
+    return () => {
+      client.disconnectUser();
+    };
+  }, [client]);
```
```suggestion
  }, [client, randomCharacter]);
```

I think some dependencies are missing here as well.

oliverlaz commented on sample-apps/react/livestream-app/src/viewers/HLSLivestream.tsx:
```diff
@@ -20,6 +20,7 @@ export const HLSLivestreamUI = () => {
   const [videoRef, setVideoRef] = useState<HTMLVideoElement | null>(null);
   const [isPlaying, setIsPlaying] = useState(false);
   useEffect(() => {
+    console.log(isBroadcasting);
```
```suggestion
```

oliverlaz commented on sample-apps/react/messenger-clone/src/App.tsx:
```diff
@@ -47,11 +47,17 @@ const Root = ({
+  );
+
+  useEffect(() => {
+    videoClient.connectUser(user, userToken);
```
```suggestion
    videoClient.connectUser(user, userToken).catch(console.error);
```

oliverlaz commented on sample-apps/react/react-video-demo/src/App.tsx:
```diff
@@ -57,14 +57,17 @@ const Init: FC<Props> = ({ incomingCallId, logo, user, token, apiKey }) => {
-    user,
-  });
+  useEffect(() => {
+    client.connectUser(user, token);
```
```suggestion
    client.connectUser(user, token).catch(console.error);
```

oliverlaz commented on sample-apps/react/zoom-clone/src/components/ChatVideoWrapper.tsx:
```diff
@@ -10,29 +7,40 @@ import type { User } from '../main';
+    if (!token || !userData) {
+      return;
+    }
+    console.log(token, userData);
```
```suggestion

```

oliverlaz reviewed (approved):
I've went through the code and I've added some suggestions around stuff that can be removed.
Also, some effects were missing dependencies, I've added them as suggestions although I'm not sure if these additions unintentionally break things :)

szuperaz commented:
> I've went through the code and I've added some suggestions around stuff that can be removed. Also, some effects were missing dependencies, I've added them as suggestions although I'm not sure if these additions unintentionally break things :)

Hm, thanks, I thought `yarn lint:all` would warn about missing dependencies for effects, as it did warn about some of them.

oliverlaz commented:
> > I've went through the code and I've added some suggestions around stuff that can be removed. Also, some effects were missing dependencies, I've added them as suggestions although I'm not sure if these additions unintentionally break things :)
> 
> Hm, thanks, I thought `yarn lint:all` would warn about missing dependencies for effects, as it did warn about some of them.

It looks like `yarn lint:all` uses some weird configuration (or, it doesn't pick up package specific configuration)...
